addons/binding/org.openhab.binding.magentatv/src/main/java/org/openhab/binding/magentatv/internal/network/MagentaTVHttp.java

```diff
@@ -0,0 +1,166 @@
+            httpResponse = HttpUtil.executeUrl(HTTP_POST, url, httpHeader, dataStream, null, NETWORK_TIMEOUT);
+            logger.trace("{} '{}' - Response = '{}'", HTTP_POST, url, httpResponse);
+            return httpResponse;
+        } catch (IOException e) {
+            String message = MessageFormat.format("Error sending HTTP POST {0} failed: {1} ({2}, response='{3}'", url,
+                    e.getMessage(), e.getClass(), httpResponse);
+            ;
+            throw new IOException(message);
+        }
+    }
+
+    /**
+     * Send raw TCP data (SUBSCRIBE command)
+     *
+     * @param remoteIp   receiver's IP
+     * @param remotePort destination port
+     * @param data       data to send
+     * @return received response
+     * @throws IOException
+     */
+    public String sendData(String remoteIp, String remotePort, String data) throws IOException {
+        String response = "";
+        Socket socket = null;
+        String errorMessage = "";
+
+        try {
+            // logger.trace("Sending data to '{}:{}': '{}'", remoteIp,
+            // remotePort, data);
+            socket = new Socket();
+            socket.setSoTimeout(4 * 1000); // set read timeout < 5s
+            socket.connect(new InetSocketAddress(remoteIp, Integer.parseInt(remotePort)), 3000);
+
+            OutputStream out = socket.getOutputStream();
+            PrintStream ps = new PrintStream(out, true);
+            ps.println(data);
+
+            InputStream in = socket.getInputStream();
+            BufferedReader buff = new BufferedReader(new InputStreamReader(in));
+
+            // wait until somthing to read is available or socket I/O fails (IOException)
+            int retry = NETWORK_TIMEOUT / 50;
+            while (!buff.ready() && retry-- > 0) {
+                Thread.sleep(50);
+            }
+            if (retry <= 0) {
+                errorMessage = MessageFormat.format("No response on {0}", data);
+            } else {
+                while (buff.ready()) {
+                    response = response + buff.readLine() + "\r\n";
+                }
+            }
+        } catch (UnknownHostException e) {
+            errorMessage = MessageFormat.format("Unknown host - {0}", e.getMessage());
+        } catch (IOException e) {
+            errorMessage = MessageFormat.format("{0} ({1})", e.getMessage(), e.getClass());
+        } catch (Exception e) {
+            errorMessage = MessageFormat.format("{0} ({1})", e.getMessage(), e.getClass());
+        } finally {
+            try {
+                if (socket != null) {
+                    socket.close();
+                }
+            } catch (Exception e) {
+            }
+        }
+
+        if (!errorMessage.isEmpty()) {
+            errorMessage = MessageFormat.format("Network I/O failed for {0}:{1}: {2}", remoteIp, remotePort,
+                    errorMessage);
+            logger.error(errorMessage);
```

Comment:
logger.error

This is also a lot of formatting for error messages. The logger will already print the origin and the exception class, so there is usually no need to prepend "Network I/O failed" for example. Just do a `logger.warn("Unknown host", e)` or `logger.warn(e);`